### DISCUSSION TITLE
Nengo_extra in nengoDL

### DISCUSSION THREAD

**Discussion by cnrgbot** [nengo/nengo-zoo#967] [Examples & Tutorials]
> _Archived from the Nengo Discourse forum._ Originally posted by **Choozi** on 2021-06-28. (Forum topic #1806, post #1 / id 5554)

Hello nengo community,

I would like to play with different neuron types that are available in nengo_extras. However, I would like to use it in nengoDL for example in the [MNIST](https://www.nengo.ai/nengo-dl/examples/spiking-mnist.html) example. Is it possible to use nengo_extras neurons in nengoDL? If no? are there any different neuron types available in nengoDL? I could not find it in the documentation.

Thank you in advance for your reply.

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **xchoo** on 2021-06-29. (Forum topic #1806, post #2 / id 5557)

Hi @​Choozi,

> *Originally @​Choozi wrote [post #1, topic #1806]:*
>
> Is it possible to use nengo_extras neurons in nengoDL?
In theory, it is possible to use the neurons from NengoExtras in a NengoDL simulation, but it would require some work to do so. You would need to re-write the neuron dynamics in TensorFlow code since NengoDL uses TensorFlow to simulate the neuron models.

Here are some forum threads that may be helpful to you if you wish to pursue this approach:
- https://forum.nengo.ai/t/adding-a-custom-neuron-type-operator-to-nengo-dl/1147
- https://forum.nengo.ai/t/my-custom-object-doesnt-work-with-nengo-dl-simulator/916

> *Originally @​Choozi wrote [post #1, topic #1806]:*
>
> are there any different neuron types available in nengoDL? I could not find it in the documentation.

Yes! NengoDL does support multiple neuron types. You can find documentation on the (default) supported neuron types [here](https://www.nengo.ai/nengo-dl/reference.html#module-nengo_dl.neurons)! NengoDL also supports all of the neurons that are available natively in Nengo (documentation [here](https://www.nengo.ai/nengo/frontend-api.html#module-nengo.neurons)).

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **Choozi** on 2021-06-29. (Forum topic #1806, post #3 / id 5560)

@​xchoo Thank you for your reply.

Thank you for sharing the useful links. 
This leads me to another question. If I am changing the neuron type it only changes the neuron type in the testing phase, right? In the training phase,  SoftLIF is used by default? am I right?

**Reply by cnrgbot** [reply]
> _Archived from the Nengo Discourse forum._ Originally posted by **xchoo** on 2021-06-29. (Forum topic #1806, post #4 / id 5572)

> *Originally @​Choozi wrote [post #3, topic #1806]:*
>
> If I am changing the neuron type it only changes the neuron type in the testing phase, right? In the training phase, SoftLIF is used by default? am I right?

It depends on how you are defining your model. NengoDL follows the TF implementation where you can specify the activation type for your layers within the model itself. If you create a model with something like this:
```
x = tf.keras.layers.Dense(1, activation=tf.nn.tanh)(...)
```
your model will be trained using the `tanh` neuron model. The one major difference between the training and testing phases (for NengoDL) is that spiking neuron types can only be use in the testing / running phase. In the training phase the rate-approximations of the spiking neurons are used.

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **Choozi** on 2021-06-30. (Forum topic #1806, post #5 / id 5576)

@​xchoo Thank you for the clarifications. :slight_smile:

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **Choozi** on 2021-07-01. (Forum topic #1806, post #6 / id 5580)

@​xchoo in the list of NengoDL neuron types we have softLIFRate and LeakyRelu. But reading their description they don't seem to be spiking neurons. In the spiking version, we have only LIF and SpikingLeakyRelu for nengoDL?  What about the neuron types that are available for nengo core [here](https://www.nengo.ai/nengo/frontend-api.html?&_ga=2.131755407.1625484257.1624895618-899412291.1624895618#module-nengo.neurons) ? can we use them in nengoDL?

**Reply by cnrgbot** [reply]
> _Archived from the Nengo Discourse forum._ Originally posted by **Eric** on 2021-07-06. (Forum topic #1806, post #7 / id 5595)

NengoDL has support for the `RegularSpiking`, `StochasticSpiking`, and `PoissonSpiking` wrappers in Nengo. These wrappers provide three different ways of turning any rate neuron into a spiking neuron. NengoDL currently supports the `LIFRate`, `RectifiedLinear`, `Sigmoid`, and `Tanh` rate neurons included in Nengo. This means you can do something like `nengo.RegularSpiking(nengo.Tanh())` to create a spiking Tanh neuron and use that in NengoDL. (Note that it's probably better to use `nengo.LIF()` rather than `nengo.RegularSpiking(nengo.LIFRate())` in NengoDL, since the former will use the SoftLIF gradient whereas the latter will not; see [this issue](https://github.com/nengo/nengo-dl/issues/143).)

Unfortunately, NengoDL does not have native TensorFlow support for `AdaptiveLIF`, `AdaptiveLIFRate`, and `Izhikevich` neurons. For these neurons, NengoDL runs them in Numpy, which works fine for the forward pass (though it is slower), but does not support differentiation (backwards pass). So if you want to use any of these in your MNIST network, you'll need to write a custom NengoDL builder that does the computation in TensorFlow (see [here](https://github.com/nengo/nengo-dl/blob/master/nengo_dl/neuron_builders.py) for the NengoDL implementations for the supported types, plus the links that @​xchoo pointed to above).

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **Choozi** on 2021-07-11. (Forum topic #1806, post #8 / id 5612)

@​Eric Thank you for your detailed reply. :slight_smile:

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **Choozi** on 2021-07-14. (Forum topic #1806, post #9 / id 5621)

@​Eric  I have been playing with these neuron types for quite a few days. I would like to study more about these wrappers' regular, stochastic and Poisson spiking. 

1) How they convert the non-spiking neuron to a spiking one. Do we have any paper references for them?

2) After training the network with the Poisson spiking wrapper, I am getting different accuracy results Evaluating the same trained network multiple times. Why is it so?


Thank you for your reply in advance.

**Reply by cnrgbot** [reply]
> _Archived from the Nengo Discourse forum._ Originally posted by **Eric** on 2021-07-14. (Forum topic #1806, post #10 / id 5623)

One place to start for reading is [A Spike in Performance](https://arxiv.org/pdf/2002.03553.pdf). It describes what goes on for `RegularSpiking`.

All three wrappers work by starting with some function/system that outputs an instantaneous firing rate at each point in time. This can be a function (e.g. ReLU, `f(x) = max(x, 0)`) where the output only depends on the input at the current point in time, or a dynamical system (e.g. `AdaptiveLIFRate` in Nengo) where the output firing rate depends on the history of inputs. That's what the "wrapped" neuron class does: it provides this instantaneous firing rate.

Given that instantaneous rate, the `PoissonSpiking` wrapper is perhaps the easiest to understand: it just uses it to sample from a [Poisson distribution], where the lambda value of the distribution is this instantaneous firing rate.

`RegularSpiking` takes the instantaneous firing rate and integrates it multiplied by a timestep `dt`. For example, if our firing rate is 5 Hz, that corresponds to one spike every 200 ms. With a timestep of 1 ms, that's one spike every 200 timesteps. `RegularSpiking` will integrate this so that one spike happens exactly every 200 timesteps given a constant instantaneous firing rate of 5 Hz. 

`StochasticSpiking` is kind of a mix between the two, and the difference mostly shows up when neurons can spike more than once per timestep. Essentially, if the instantaneous firing rate says we should have e.g. 2.4 spikes per timestep, than on a given timestep we'd always have at least 2 spikes, and a 0.4 probability (40% chance) of a third spike.

The reason you get different results with `PoissonSpiking` is because it's stochastic, so it will output different spike trains each time. You can set the seed on the [nengo.Simulator](https://www.nengo.ai/nengo/backend-api.html#nengo.Simulator) to seed this randomness so it's the same each time.

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **Choozi** on 2021-07-15. (Forum topic #1806, post #11 / id 5642)

@​Eric thank you for your answers. :slight_smile: 

I have found one conversion example explained in [here](https://forum.nengo.ai/t/converting-keras-model-to-spiking-with-tanh-activation-function/1506/2). However, what is this gain parameter? (you set it to 3 for Tanh in the example).

Normally the conventional Tanh starts from -1 to 1. The spiking one starts from 0 and then for few time steps it is 0 and then it starts going down towards -400 and starts behaving like Tanh in a spiking way. I this because it starts accumulating current in the start?

Now how would an input value be mapped? How would this mapping work? For instance, I have an input value of 1 it would be mapped to approx. to 0.77 in conventional Tanh. What about in spiking version? How would this mapping look like? 

I am sorry if this question has been already answered on the forum.

Thank you for your answers.

**Reply by cnrgbot** [reply]
> _Archived from the Nengo Discourse forum._ Originally posted by **xchoo** on 2021-07-16. (Forum topic #1806, post #12 / id 5656)

> *Originally @​Choozi wrote [post #11, topic #1806]:*
>
> However, what is this gain parameter? (you set it to 3 for Tanh in the example).

If you plot the `tanh` function, you'll notice that the slope of the curve at the midpoint is roughly 45 degrees. The `gain` parameter here changes this slope, and having a bigger gain value increases the "steepness" of this slope. You can experiment with it yourself by changing the `gain` in the code to see what effect it has. :smiley: 

> *Originally @​Choozi wrote [post #11, topic #1806]:*
>
> Normally the conventional Tanh starts from -1 to 1. The spiking one starts from 0 and then for few time steps it is 0 and then it starts going down towards -400 and starts behaving like Tanh in a spiking way. I this because it starts accumulating current in the start?

That is essentially correct. In Nengo, if you do not specify any defaults, the neurons will start with a blank state (i.e., 0 input current, 0 membrane current, etc.). This means that when the simulation starts, the neurons aren't doing anything (and outputs no spikes), and it takes a few timesteps for the neuron to get going. We refer to these as "startup transients".

> *Originally @​Choozi wrote [post #11, topic #1806]:*
>
> Now how would an input value be mapped? How would this mapping work? For instance, I have an input value of 1 it would be mapped to approx. to 0.77 in conventional Tanh. What about in spiking version? How would this mapping look like?

The mapping would be similar in the spiking version, with the addition that the "output value" is scaled by the firing rate. As an example, if you had a "conventional" `tanh` function (i.e., the `gain=1`), an input value of 1 would have an output value of 0.77. In the spiking version, the output would be this value multiplied by the maximum firing rate. In @​Eric's example code, the maximum firing rate of the `tanh` neuron is 400Hz, so the output of such a neuron with an input of 1, would be $400 \times 0.77 \approx 308$. 

If you change the `gain` in the code to `1`, and run Eric's code, you will see a graph like this:
![image](https://raw.githubusercontent.com/nengo/forum-archive/main/uploads/default/original/1X/f9a9144fa2023ab197e97caec07b7d2e3bbc1800.png)

And it indeed shows that when the input to the neuron is 1 (i.e., at `t=2s`), the output firing rate is about 300.

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **Choozi** on 2021-07-16. (Forum topic #1806, post #13 / id 5661)

Thank you @​xchoo for your detailed reply.

> The `gain` parameter here changes this slope, and having a bigger gain value increases the “steepness” of this slope. You can experiment with it yourself by changing the `gain` in the code to see what effect it has. :smiley:

Does this mean that we can somehow control the response of the neuron by adjusting this gain parameter? 
My question is then when I am using tanh in nengoDL then this gain parameter is learned by the network during training?

Thank for the rest of the explanations. :-) 

But I may sound stupid, but just want to clarify this tanh and others non-spiking neurons are simulated one  or built with some mathematical formulations?

**Reply by cnrgbot** [reply]
> _Archived from the Nengo Discourse forum._ Originally posted by **xchoo** on 2021-07-17. (Forum topic #1806, post #14 / id 5664)

> *Originally @​Choozi wrote [post #13, topic #1806]:*
>
> Does this mean that we can somehow control the response of the neuron by adjusting this gain parameter?

In some respects yes. You can definitely change the behaviour of the neuron by changing the `gain` value, that's why it is made available to you in the API.

> *Originally @​Choozi wrote [post #13, topic #1806]:*
>
> My question is then when I am using tanh in nengoDL then this gain parameter is learned by the network during training?

NengoDL doesn't allow you to train the gains directly. The gain of the neuron is contained within the connection weights between the layers and by training those weights, you indirectly train the gains. 

> *Originally @​Choozi wrote [post #13, topic #1806]:*
>
> But I may sound stupid, but just want to clarify this tanh and others non-spiking neurons are simulated one or built with some mathematical formulations?

I'm not exactly sure what you are asking here? What do you mean by "simulated" as opposed to "built with a mathematical formulation"? All of the neurons in Nengo are simulated, which is to say that their behaviour is the result of evaluating a mathematical formula (that describes the neuron) over a period of time.

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **Choozi** on 2021-08-11. (Forum topic #1806, post #15 / id 5727)

@​xchoo 
Thank you very much for your answer and for the clarifications. :) This is what I wanted to know :)
